plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java

```diff
@@ -23,14 +25,73 @@
  * Validates a list of {@link GModelElement}s based on a set of validation rules and returns corresponding issue
  * {@link Marker}s.
  * An issue marker is a serializable description of the validation violation that can be visualised by the GLSP client.
+ *
+ * There are two default reasons for a validation:
+ * <ol>
+ * <li><i>live</i> validation is executed on open and after each operation. These validation rules should be rather
+ * fast. Overwrite {@link #doLiveValidation(GModelElement)} to implement the live validation rules.</li>
+ * <li><i>batch</i> validation is executed on demand by the client. These validation rules can be more expensive.
+ * Overwrite {@link #doBatchValidation(GModelElement)} to implement the batch validation rules.</li>
+ * </ol>
  */
 public interface ModelValidator {
 
    /**
     * Validates the given list of {@link GModelElement}s and returns a list of {@link Marker}s.
     *
     * @param elements The list of {@link GModelElement} to validate.
+    * @param reason   The reason for a validation request, such as "batch" or "live" validation.
     * @return A list of {@link Marker}s for the validated {@link GModelElement}s.
     */
-   List<Marker> validate(GModelElement... elements);
+   default List<Marker> validate(final List<GModelElement> elements, final String reason) {
+      List<Marker> markers = new ArrayList<>();
+
+      for (GModelElement element : elements) {
+         if (MarkersReason.LIVE.equals(reason)) {
```

Comment:
What is the assumption here regarding the reason property? Is it a fixed string that can only either be `batch` or `live`? Or can adopters also introduce custom reasons? If so we should probably explicitly check for the `batch` reason as well and add a fallback case for unsupported reasons.

Reply:
Right, the assumption is that we use batch or live by default, but adopters may add custom reasons, e.g. some specific analysis and use custom strings here. My thinking was that they then would have to overwrite the `validate()` method. But you are right, it'd be more convenient to handle the batch case explicitly and have a `default doValidationForCustomReason()` that needs to be overwritten in those cases.